common/dynamicconfig/collection.go

```diff
@@ -81,10 +113,86 @@ var (
 
 // NewCollection creates a new collection
 func NewCollection(client Client, logger log.Logger) *Collection {
-	return &Collection{
-		client:   client,
-		logger:   logger,
-		errCount: -1,
+	c := &Collection{
+		client:        client,
+		logger:        logger,
+		errCount:      -1,
+		subscriptions: make(map[Key]map[int]any),
+	}
+	s := DynamicConfigSubscriptionCallback.Get(c)()
+	c.callbackPool = goro.NewAdaptivePool(clock.NewRealTimeSource(), s.MinWorkers, s.MaxWorkers, s.TargetDelay, s.ShrinkFactor)
+	if subcli, ok := client.(SubscribableClient); ok {
+		c.cancelClientSubscription = subcli.Subscribe(c.keysChanged)
+	} else {
+		c.poller.Go(c.pollForChanges)
+	}
+	return c
+}
+
+func (c *Collection) Stop() {
+	c.poller.Cancel()
+	c.poller.Wait()
+	if c.cancelClientSubscription != nil {
+		c.cancelClientSubscription()
+	}
+	c.subscriptionLock.Lock()
+	defer c.subscriptionLock.Unlock()
+	c.callbackPool.Stop()
+}
+
+// Implement pingable.Pingable
+func (c *Collection) GetPingChecks() []pingable.Check {
+	return []pingable.Check{
+		{
+			Name:    "dynamic config callbacks",
+			Timeout: 5 * time.Second,
+			Ping: func() []pingable.Pingable {
+				var wg sync.WaitGroup
+				wg.Add(1)
+				c.callbackPool.Do(wg.Done)
+				wg.Wait()
+				return nil
+			},
+		},
+	}
+}
+
+func (c *Collection) pollForChanges(ctx context.Context) error {
+	for ctx.Err() == nil {
+		util.InterruptibleSleep(ctx, DynamicConfigSubscriptionPollInterval.Get(c)())
```

Comment:
sure. except maybe I'll do it just as a local var in this function? this is the fallback implementation that's not used unless someone overrides the dynamic config client so it would be a (tiny) waste to do it always.